File: hashTable/longestConsecutiveSequence.py
def longestConsecutiveSequence(nums):
    # edge case if no nums
    if len(nums) == 0:
        return 0
    # sort the nums list
    nums.sort()
    print('nums', nums)
    # max = 0
    maxLength = 0

    # for num in list
    for num in nums:
        # if num - 1 in list, continue (skip)
        if num - 1 in nums:
            continue

        # currNum = num
        currNum = num
        # currMax = 1
        currMax = 1

        # while currNum + 1 in list
        while currNum + 1 in nums:
            # currMax += 1
            currMax += 1
            # currNum += 1
            currNum += 1

        # max = max(max, currMax)
        maxLength = max(maxLength, currMax)

    # return max
    return maxLength


numbers = [100, 4, 200, 1, 3, 2]

print(longestConsecutiveSequence(numbers))

File: hashTable/longestConsecutiveSequence.js
// Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.
// You must write an algorithm that runs in O(n) time.

// Example 1:

// Input: nums = [100,4,200,1,3,2]
// Output: 4
// Explanation: The longest consecutive elements sequence is [1, 2, 3, 4]. Therefore its length is 4.
// Example 2:

// Input: nums = [0,3,7,2,5,8,4,6,0,1]
// Output: 9

// clean solution found on Leetcode:
function longestConsecutive(nums) {
  if (nums == null || nums.length === 0) return 0;

  const set = new Set(nums);
  let max = 0;

  for (let num of set) {
    if (set.has(num - 1)) continue; // make sure starting from the beginning of sequence - we are going to be counting up from a starting number so we need to make sure nothing comes before it.

    let currNum = num;
    let currMax = 1;

    while (set.has(currNum + 1)) {
      currNum++;
      currMax++;
    }
    max = Math.max(max, currMax);
  }

  return max;
}

console.log(longestConsecutive([100, 4, 200, 1, 3, 2]));
